### ISSUE TITLE
Allow for RecoveryCallback to throw original exception

### ISSUE THREAD

**Issue by shollander** [spring-projects/spring-retry#135]
The current design of the RetryTemplate and RecoveryCallback is to not throw any exception if a recovery callback is provided. There are cases where even after recovery, the caller wants to know if the original action failed or succeeded. This can be accomplished by providing a way for the recovery callback to throw the original exception. It is currently a little clunky to do so since RecoveryCallback has a signature of `throws Exception`, while `retryContext.getLastThrowable()` provides a `Throwable`, which is not compatible.

**Comment by dsyer** [comment]
I guess we could change the signature if we are going to bump to a new major version. Would that solve your problem? Under what circumstances would you care? Why is `Exception` not broad enough?

**Comment by shollander** [comment]
Because this will not compile:
```
RecoveryCallback<T> = retryContext -> {
    throw retryContext.getLastThrowable();
};
```

**Comment by dsyer** [comment]
Fair enough. But it's easy to work around isn't it? I guess that's what you meant by "clunky".

**Comment by dsyer** [comment]
Sure, but you would put those 6 lines in a convenience class and only use one line in practice, if this pattern was common. We can put such a convenience class in spring-retry if you like, but there are plenty already out there. Plus it’s easy to write your own.

**Comment by shollander** [comment]
Granted, although that doesn't address the second issue of unnecessarily wrapping the original exception in certain cases.

**Comment by dsyer** [comment]
Do you have an actual use case where that matters?

**Comment by shollander** [comment]
Not at the moment. My point was just that the API is slightly inconsistent here and can cause inconveniences. If you don't want to fix it there are workarounds.

**Comment by artembilan** [comment]
The project goes into a maintenance mode, therefore all users should consider to migrate to Spring Framework Retry API instead: https://spring.io/blog/2025/09/09/core-spring-resilience-features.

There is no `recover` abstraction any more.
The user is free to catch `RetryException` and analyze it for the further logic.
